core/vibes/soul/primitives/product-card/compare.tsx

```diff
@@ -1,38 +1,57 @@
 'use client';
 
 import { parseAsArrayOf, parseAsString, useQueryState } from 'nuqs';
+import { startTransition } from 'react';
 
 import { Checkbox } from '@/vibes/soul/form/checkbox';
 
+import { useCompareDrawer } from '../compare-drawer';
+
+interface CompareDrawerItem {
+  id: string;
+  image?: { src: string; alt: string };
+  href: string;
+  title: string;
+}
+
 interface Props {
-  productId: string;
   colorScheme?: 'light' | 'dark';
   paramName?: string;
   label?: string;
+  product: CompareDrawerItem;
 }
 
 export const Compare = function Compare({
-  productId,
   colorScheme = 'light',
   paramName = 'compare',
   label = 'Compare',
+  product,
 }: Props) {
-  const [param, setParam] = useQueryState(
+  const [, setParam] = useQueryState(
     paramName,
     parseAsArrayOf(parseAsString).withOptions({ shallow: false }),
   );
 
+  const { optimisticItems, setOptimisticItems } = useCompareDrawer();
+
   return (
     <Checkbox
-      checked={param?.includes(productId) ?? false}
+      checked={!!optimisticItems.find((item) => item.id === product.id)}
       colorScheme={colorScheme}
       label={label}
       onCheckedChange={(value) => {
         void setParam((prev) => {
           const next =
             value === true
-              ? [...(prev ?? []), productId]
-              : (prev ?? []).filter((v) => v !== productId);
+              ? [...(prev ?? []), product.id]
+              : (prev ?? []).filter((v) => v !== product.id);
+
+          startTransition(() => {
+            setOptimisticItems({
+              type: value === true ? 'add' : 'remove',
+              item: product,
+            });
+          });
 
           return next.length > 0 ? next : null;
         });
```

Comment:
`setParam` itself should be inside the transition so we can `await` it instead of using `void`. And ideally the optimistic action happens before that

Reply:
Fixed.